src/redux/action-creators/token/transfer.ts

```diff
@@ -1,65 +1,103 @@
+
+// Actions
+const tokenTransferRequest = (value: Transfer): FSA => ({
+  type: TOKEN_TRANSFER_REQUEST,
+  payload: value,
+})
+
+const tokenTransferOk = (value: Transfer): FSA => ({
+  type: TOKEN_TRANSFER_OK,
+  payload: value,
+})
+
+const tokenTransferError = (value: Error): FSA => ({
+  type: TOKEN_TRANSFER_ERROR,
+  payload: value,
+})
+
+const tokenTransferReset = (): FSA => ({
+  type: TOKEN_TRANSFER_RESET,
+  payload: {},
+})
+
+// Action Creators
 
 // TODO type the returned thunk
-const transfer = (to:string, amount:Nos, from:string): any => {
+const transfer = (to: string, amount: number | string, from?: string): any =>
   // TODO type the thunk args
-  return async (dispatch:any, getState:any): Promise<TransactionReceipt|null> => {
-    const state:State = getState(),
-      // a token must have been deployed by this point
-      tokenAddress = addressSelector(state),
-      ws = state.websocketAddress || '',
-      web3 = new Web3(new Web3.providers.WebsocketProvider(ws)),
-      contract = new Erc20(from)
+  async (dispatch: Function, getState: Function): Promise<{ [key: string]: string } | undefined> => {
+    const state: State = getState()
+    const owner = getOwner(state)
+
+    // a token must have been deployed by this point
+    const tokenAddress = addressSelector(state)
+    const ws = state.websocketAddress || ''
+    const web3Provider = new Web3.providers.WebsocketProvider(ws)
+    const web3 = new Web3(web3Provider)
+    const contract = new Erc20(owner.address)
+
     // instantiate a contract from the deployed token
-    tokenAddress && await contract.at(web3, { address: tokenAddress })
-
-    let tx = null
-
-    if (!contract) dispatch(transferError(new Error(Errors.NO_TOKEN_FOUND)))
-    else {
-      // dispatch transfer early as an in-flight notification
-      dispatch(transferral(TRANSFER, to, amount, from))
-      // try the actual on-chain transfer
-      try {
-        // we can allow the contract to fallback on the default account it was made from
-        tx = await contract.transfer(to, amount)
-        dispatch(transferral(TRANSFERRED, to, amount, from))
-      } catch(err) {
-        dispatch(transferError(err))
-      }
+    tokenAddress && await contract.at(web3, { address: tokenAddress }, { from: owner.address })
+
+    if (!contract) {
+      const error = new Error(Errors.NO_TOKEN_FOUND)
+      dispatch(tokenTransferError(error))
+
+      return undefined
     }
 
-    return tx
+    const args: Transfer = { to, amount, from: from || owner.address }
+
+    // dispatch that a request has been initialized
+    dispatch(tokenTransferRequest(args))
+
+    // try the actual on-chain transfer
+    try {
+      const emitter = contract.getEventEmitter('Transfer')
+
+      // we can allow the contract to fallback on the default account it was made from
+      contract.transfer(to, amount)
+
+      const eventLog: EventLog = await onData(emitter)
+      const eventValues = eventLog.returnValues
+
```

Comment:
i like it. we'll make a ticket for updating `eventReturnValues` like i mentioned before... this is good for now tho